File: code_snippets/ts/src/use_cases/async_tasks/sync_to_async/file_upload_workflow.ts
import * as restate from "@restatedev/restate-sdk";
import {
  WorkflowContext,
  WorkflowSharedContext,
} from "@restatedev/restate-sdk";

// <start_here>
const fileUploadWorfklow = restate.workflow({
  name: "FileUploadWorfklow",
  handlers: {
    run: async (ctx: WorkflowContext) => {
      // <mark_1>
      const url = await ctx.run(() => createS3Bucket());
      await ctx.run(() => uploadFile(url));

      // <mark_2>
      await ctx.promise<URL>("url").resolve(url);
      // </mark_2>
      
      return url;
      // </mark_1>
    },

    getUrlViaEmail: async (
      ctx: WorkflowSharedContext,
      req: { email: string }
    ) => {
      // <mark_2>
      const url = await ctx.promise<URL>("url");
      // </mark_2>
      await ctx.run(() => sendEmail(url, req.email));
    },
  },
});

export type FileUploadWorkflow = typeof fileUploadWorfklow;
// <end_here>

async function createS3Bucket(): Promise<URL> {
  const bucket = Number(Math.random() * 1_000_000_000).toString(16);
  return new URL(`https://s3-eu-central-1.amazonaws.com/${bucket}/`);
}

async function uploadFile(target: URL) {
  // simulate some work by delaying for a while. sometimes takes really long.
  return new Promise((resolve) =>
    setTimeout(resolve, Math.random() < 0.0 ? 1_500 : 10_000)
  );
}

async function sendEmail(url: URL, email: string) {
  // send email
  console.log(` >>> Sending email to '${email}' with URL ${url}`);
}

File: code_snippets/python/src/use_cases/async_tasks/sync_to_async/file_upload_workflow.py
from restate import WorkflowContext, WorkflowSharedContext, Workflow
from typing import TypedDict


class FileUploadWorkflow(TypedDict):
    name: str
    handlers: dict


def create_s3_bucket() -> str:
    return ""


def upload_file(target: str):
    pass


def send_email(url: str, email: str):
    print(f" >>> Sending email to '{email}' with URL {url}")


# <start_here>
file_upload_workflow = Workflow("FileUploadWorkflow")


@file_upload_workflow.main()
async def run(ctx: WorkflowContext) -> str:
    url = await ctx.run("bucket creation", lambda: create_s3_bucket())
    await ctx.run("upload", lambda: upload_file(url))

    # <mark_1>
    await ctx.promise("url").resolve(url)
    # </mark_1>

    return url


@file_upload_workflow.handler("getUrlViaEmail")
async def get_url_via_email(ctx: WorkflowSharedContext, email: str):
    # <mark_1>
    url = await ctx.promise("url").value()
    # </mark_1>
    await ctx.run("email", lambda: send_email(url, email))
# <end_here>
